File: android_mobile/src/main/java/com/alexstyl/specialdates/search/SearchResultsAdapter.java
package com.alexstyl.specialdates.search;

import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.ViewGroup;

import com.alexstyl.specialdates.contact.Contact;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.date.DateLabelCreator;
import com.alexstyl.specialdates.events.namedays.NameCelebrations;
import com.alexstyl.specialdates.images.ImageLoader;
import com.novoda.notils.exception.DeveloperError;

import java.util.ArrayList;
import java.util.List;

public final class SearchResultsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final List<ContactEventViewModel> searchResults = new ArrayList<>();

    private NamedayCard namedayCard = new NamedayCard();

    private boolean isLoadingMore;
    private boolean canLoadMore = false;
    private final ImageLoader imageLoader;
    private final DateLabelCreator labelCreator;

    SearchResultsAdapter(ImageLoader imageLoader, DateLabelCreator labelCreator) {
        this.imageLoader = imageLoader;
        this.labelCreator = labelCreator;
    }

    void updateSearchResults(SearchResults searchResults) {
        this.searchResults.clear();
        this.searchResults.addAll(searchResults.getViewModels());

        if (this.canLoadMore != searchResults.getCanLoadMore()) {
            this.canLoadMore = searchResults.getCanLoadMore();
        }
        notifyDataSetChanged();
    }

    public void setNamedays(NameCelebrations namedays) {
        namedayCard.setNameday(namedays);
        notifyDataSetChanged();
    }

    void notifyIsLoadingMore() {
        isLoadingMore = true;
        notifyDataSetChanged();
    }

    void clearResults() {
        this.searchResults.clear();
        this.namedayCard.clear();
        this.isLoadingMore = false;
        this.canLoadMore = false;
        notifyDataSetChanged();
    }

    interface SearchResultClickListener {

        void onContactClicked(Contact contact);

        void onNamedayClicked(Date date);
    }

    private SearchResultClickListener listener;

    void setSearchResultClickListener(SearchResultClickListener l) {
        this.listener = l;
    }

    private static final int VIEWTYPE_CONTACTVIEW = 0;
    private static final int VIEWTYPE_NAMEDAYS_VIEW = 1;
    private static final int VIEWTYPE_LOAD_MORE = 2;
    private static final int VIEWTYPE_NO_RESULTS = 3;

    @Override
    public int getItemViewType(int position) {

        if (loadMoreCardAt(position)) {
            return VIEWTYPE_LOAD_MORE;
        }
        if (namedayCardAt(position)) {
            return VIEWTYPE_NAMEDAYS_VIEW;
        }
        if (containsNoResults()) {
            return VIEWTYPE_NO_RESULTS;
        }
        return VIEWTYPE_CONTACTVIEW;
    }

    private boolean loadMoreCardAt(int position) {
        return canLoadMore && position == getItemCount() - 1;
    }

    private boolean namedayCardAt(int position) {
        return namedayCard.isAvailable() && position == 0;
    }

    private boolean containsNoResults() {
        return searchResults.size() == 0 && !namedayCard.isAvailable();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        if (viewType == VIEWTYPE_CONTACTVIEW) {
            return SearchResultContactViewHolder.createFor(parent, imageLoader);
        }
        if (viewType == VIEWTYPE_NAMEDAYS_VIEW) {
            return SearchResultNamedayViewHolder.createFor(parent, labelCreator);
        }

        if (viewType == VIEWTYPE_LOAD_MORE) {
            return MoreViewHolder.createFor(parent);
        }
        if (viewType == VIEWTYPE_NO_RESULTS) {
            return NoResultsViewHolder.createFor(parent);
        }

        throw new DeveloperError("No Holder created for type " + viewType);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder vh, int position) {

        int type = getItemViewType(position);
        if (type == VIEWTYPE_CONTACTVIEW) {
            if (isDisplayingNamedayCard()) {
                // nameday adds one more row
                position = position - 1;
            }
            ContactEventViewModel viewModel = searchResults.get(position);
            SearchResultContactViewHolder viewHolder = (SearchResultContactViewHolder) vh;
            viewHolder.bind(viewModel, listener);
        } else if (type == VIEWTYPE_NAMEDAYS_VIEW) {
            ((SearchResultNamedayViewHolder) vh).bind(namedayCard, listener);
        } else if (type == VIEWTYPE_LOAD_MORE) {
            ((MoreViewHolder) vh).bind(isLoadingMore);
        } else if (type == VIEWTYPE_NO_RESULTS) {
            // no special binding needed
        } else {
            throw new DeveloperError("Unhandled type " + type);
        }

    }

    private boolean isDisplayingNamedayCard() {
        return namedayCard.isAvailable();
    }

    @Override
    public int getItemCount() {
        // TODO move all this to presenter's viewmodel logic
        return searchResults.size()
                + (namedayCard.isAvailable() ? 1 : 0)
                + (canLoadMore ? 1 : 0)
                + (shouldDisplayEmptyRow() ? 1 : 0);
    }

    private boolean shouldDisplayEmptyRow() {
        return hasSearchQuery() && (!canLoadMore && containsNoResults());
    }

    private boolean hasSearchQuery() {
        return !TextUtils.isEmpty("");
    }

}

File: android_mobile/src/main/java/com/alexstyl/specialdates/search/SearchResultsAdapter.kt
package com.alexstyl.specialdates.search

import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import com.alexstyl.specialdates.R
import com.alexstyl.specialdates.images.ImageLoader
import com.alexstyl.specialdates.search.SearchResultViewHolder.ContactReadPermissionRequestViewHolder
import com.alexstyl.specialdates.search.SearchResultViewHolder.SearchResultContactViewHolder
import com.alexstyl.specialdates.search.SearchResultViewHolder.SearchResultNamedayViewHolder
import com.alexstyl.specialdates.search.SearchResultViewModel.ContactReadPermissionRequestViewModel
import com.alexstyl.specialdates.search.SearchResultViewModel.ContactSearchResultViewModel
import com.alexstyl.specialdates.search.SearchResultViewModel.NamedaySearchResultViewModel
import com.alexstyl.specialdates.ui.widget.ColorImageView
import java.lang.IllegalStateException

class SearchResultsAdapter(private val imageLoader: ImageLoader,
                           private val listener: SearchResultClickListener)
    : RecyclerView.Adapter<SearchResultViewHolder<SearchResultViewModel>>() {

    private val searchResults = mutableListOf<SearchResultViewModel>()

    override fun getItemViewType(position: Int): Int {
        val viewModel = searchResults[position]
        return when (viewModel) {
            is NamedaySearchResultViewModel -> VIEWTYPE_NAMEDAY
            is ContactSearchResultViewModel -> VIEWTYPE_CONTACT
            is ContactReadPermissionRequestViewModel -> VIEWTYPE_PERMISSION
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SearchResultViewHolder<SearchResultViewModel> = when (viewType) {
        VIEWTYPE_CONTACT -> contactViewHolder(parent)
        VIEWTYPE_NAMEDAY -> namedayViewHolder(parent)
        VIEWTYPE_PERMISSION -> permissionRequestViewHolder(parent)
        else -> {
            throw IllegalStateException("Cannot create ViewHolder for viewType [$parent]")
        }
    }

    private fun contactViewHolder(parent: ViewGroup): SearchResultViewHolder<SearchResultViewModel> {
        val layoutInflater = LayoutInflater.from(parent.context)
        val view = layoutInflater.inflate(R.layout.row_search_result_contact_event, parent, false)
        val nameView = view.findViewById(R.id.search_result_contact_name) as TextView
        val avatarView = view.findViewById(R.id.search_result_avatar) as ColorImageView
        return SearchResultContactViewHolder(view, avatarView, nameView, imageLoader)
                as SearchResultViewHolder<SearchResultViewModel> // <- TODO how to get rid of this?
    }

    private fun namedayViewHolder(parent: ViewGroup): SearchResultViewHolder<SearchResultViewModel> {
        val layoutInflater = LayoutInflater.from(parent.context)
        val view = layoutInflater.inflate(R.layout.row_search_result_nameday, parent, false)
        val namedayView = view.findViewById(R.id.name_celebrating) as TextView
        val datesLayout = view.findViewById(R.id.dates) as LinearLayout
        return SearchResultNamedayViewHolder(view, namedayView, datesLayout)
                as SearchResultViewHolder<SearchResultViewModel>// <- TODO how to get rid of this?
    }

    private fun permissionRequestViewHolder(parent: ViewGroup): SearchResultViewHolder<SearchResultViewModel> {
        val layoutInflater = LayoutInflater.from(parent.context)
        val view = layoutInflater.inflate(R.layout.row_search_result_permission_needed, parent, false)
        return ContactReadPermissionRequestViewHolder(view)
                as SearchResultViewHolder<SearchResultViewModel> // <- TODO how to get rid of this?
    }

    override fun onBindViewHolder(holder: SearchResultViewHolder<SearchResultViewModel>, position: Int) {
        val viewModel = searchResults[position]
        holder.bind(viewModel, listener)
    }

    override fun getItemCount() = searchResults.size

    fun displaySearchResults(searchResults: List<SearchResultViewModel>) {
        val diffCallback = SearchResultDiffCallback(this.searchResults, searchResults)
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        this.searchResults.clear()
        this.searchResults.addAll(searchResults)
        diffResult.dispatchUpdatesTo(this)
    }

    companion object {
        private const val VIEWTYPE_CONTACT = 0
        private const val VIEWTYPE_NAMEDAY = 1
        private const val VIEWTYPE_PERMISSION = 2
    }

}
